text/0000-remove-shell-processes.md

```diff
@@ -0,0 +1,401 @@
+
+**complexity** - Having a direct dependency between the launcher and specific shells is inelegant. It makes the spec more complex and end users and buildpack author are forced to understand that complexity in order to understand the behavior of the resulting usage or debug issues. For example end users and buildpacks authors must understand:
+- The difference between a direct and shell process
+- The nuances of argument handling in direct vs. Bash vs. Cmd cases
+- That buildpack-provided profile scripts will not apply to direct processes
+- That a user-provided `.profile` script will not apply to a direct processes
+
+ Removing this special behavior and requiring buildpacks to explicitly include any required shell in the command itself creates a simpler, more comprehensible model for buildpack authors and end users alike.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Process Schema
+The existing schema for the processes table in `launch.toml` is
+```
+# Old Schema
+
+[[processes]]
+type = "required"
+command = "required"
+args = ["optional"]
+direct = false
+default = false
+```
+
+The new proposed schema is:
+```
+# New Schema
+
+[[processes]]
+type = "required"
+command = ["required"]
+args = ["optional"]
+default = false
+```
+The following changes have been made:
+- **`direct` has been removed** - all processes are executed directly. If `bash` or `cmd.exe` is required it should be included in `command`. No surprises.
+- **`command` is now an array** - Arguments in `command` *will not* be overwritten if a user provides additional arguments at runtime. `args` are default arguments that *will* be overwritten if a user provides additional arguments at runtime. The makes `command` analogous to `Entrypoint` in the OCI spec and `command` in a Kubernetes PodSpec. `args` is analogous to `cmd` and `args` in docker and Kubernetes respectively.
+
+## Using Environment Variables in a Process
+
+One upside to our previous execution strategy was that it enable users to include environment variable references in arguments that were later evaluated in the container. To preserve this feature we can instead adopt a variabtion on the Kubernetes strategy for [environment variables interpolation](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/#using-environment-variables-inside-of-your-config). If a buildpack or user includes `${<env>}` in the `command` or `args` and `{env}` is the name of an environment variable set in the launch environment, the launcher will replace this string with the value of the environment variable after apply buildpack-provided env modifications and before launching the process.
```

Comment:
Is there a risk that in doing this we might somehow expose secrets? For example if the process type is:

```
./bin/server --key $MY_SECRET
```

If we substitute `MY_SECRET` in the process could it end up in the logs or somewhere that it might not when Bash substitutes it? 

Are they known libraries/mechanisms we can use to make sure we're on the right-side of the law here?

Reply:
I think we are right that there is some cause for concern here. I think we should change this RFC and switch the interpolation notation to something like `$((MY_SECRET))` so we never interpolate environment variables that are intended for Bash to interpolate in a command like `["bash", "-c", "./bin/server --key ${MY_SECRET}"]. It's a little sad that we need yet another notation but it seems good to remove any ambiguity with regards to what gets interpolated by k8s, vs the launcher, vs Bash (if it is explicitly included in a command).

In this example are you imagining that it is the user you provides `$MY_SECRET` dynamically at runtime? I would argue that this is not a safe thing to do with secrets because they end up in the container config. Secrets should be provided via a safe mechanism like bindings and buildpacks should be responsibility for protecting the secret, which hopefully will not require passing it directly on the command line (even as a reference for Bash to interpolate) in the vast majority of use cases.

Reply:
`$(ENV)` should work? (the k8s convention). It only works in k8s for env vars directly defined  in the spec, the rest are left untouched.

Reply:
updated to use k8s style notation in https://github.com/buildpacks/rfcs/pull/168/commits/827cdd5782eceed7ba367941af03dd8d519bb3d3